src/domain/repository/BookCopyRepositoryImpl.java

```diff
@@ -0,0 +1,133 @@
+import exception.ValidationException;
+import service.DatabaseUtility;
+
+import java.awt.print.Book;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BookCopyRepositoryImpl implements BookCopyRepository {
+
+    /**
+     /***
+     *
+     ID BIGINT NOT NULL  AUTO_INCREMENT, -- BOOK COPY ID
+     BOOK_ID BIGINT NOT NULL, -- the book id. Foreign key to book table(ben)
+     */
+
+    @Override
+    public BookCopy findById(long bookCopyId) {
+        String query = "SELECT ID, BOOK_ID FROM BOOK_COPY  WHERE ID = ? ";
+
+        try {
+            PreparedStatement ps = DatabaseUtility.getConnection().prepareStatement(query);
+            ps.setLong(1, bookCopyId);
+            ResultSet rs = ps.executeQuery();
+            if (rs.next()) {
+                BookCopy b = new BookCopy();
+                b.setId(rs.getLong("ID"));
+                b.setBookId(rs.getLong("BOOK_ID"));
+                return b;
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+            throw new SystemException("Exception Happened while trying to find book copy ID "+ bookCopyId);
+        } finally {
+            DatabaseUtility.releaseConnection();
+        }
+
+        return null;
+    }
+
+    @Override
+    public List<BookCopy> findByBookId(long bookId) {
+        String query = "SELECT ID, BOOK_ID FROM BOOK_COPY  WHERE BOOK_ID = ? ";
+
+        try {
+            PreparedStatement ps = DatabaseUtility.getConnection().prepareStatement(query);
+            ps.setLong(1, bookId);
+            ResultSet rs = ps.executeQuery();
+            List<BookCopy> books = new ArrayList<>();
+
+            while (rs.next()) {
+                BookCopy bookCopy = new BookCopy();
+                bookCopy.setId(rs.getLong("ID"));
+                bookCopy.setBookId(rs.getLong("BOOK_ID"));
+                books.add(bookCopy);
+            }
+            return books;
+        } catch (SQLException e) {
+            e.printStackTrace();
+            throw new SystemException("Exception Happened while trying to find book ID "+ bookId);
+        } finally {
+            DatabaseUtility.releaseConnection();
+        }
+    }
+
+    @Override
+    public BookCopy save(BookCopy bookCopy) throws ValidationException {
+        if (bookCopy == null){
+            throw new ValidationException("Unable to save a null book. Please add a book.",
+                    new ValidationError("Book", "null. Please enter value", "Can't Save a null book"));
+        }
+
+        BookCopy dbBookCopy  = findById(bookCopy.getId());
+        if (dbBookCopy != null){
+            return dbBookCopy;
+        }
+        else {
+            return (insert(bookCopy));
+        }
+    }
+
+    private BookCopy insert(BookCopy bookCopy) {
+        String query = "INSERT INTO BOOK_COPY (BOOK_ID, ID)  VALUES (?, ?)";
+
+        try {
```

Comment:
you need to make sure that the book exists. otherwise the foreign key will not allow you to insert into the bookcopy table.

Reply:
Unsure as to how to do this. Deleted the taking in of the ID as that is assigned by auto-increment. For foreign key, I will still use .getBookId, however I had needed to say that t is a foreign key which was done in the creation of the table. How do I use the foreign key from Book Table. Since we only have id, and bookId how will you correlate the id foreign key and the id in Book_Table. Will you need to give the bookId in the class name?

Reply:
if (bookCopy.getBookId() == 0L){
            throw new SystemException("Cannot pass null ")
        } 
added this though I am unsure if it is significant.